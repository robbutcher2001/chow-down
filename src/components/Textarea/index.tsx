import React, { Component, ChangeEvent } from 'react';

import styled from 'styled-components';
import { Fields, FieldValidations } from '../Form';

export interface TextareaProps {
  name: string,
  label: string,
  validator: (value: string) => boolean,
  form?: Fields,
  validFields?: FieldValidations,
  setNewFormState?: (field: string, newValue: string) => void,
  setValidationState?: (field: string, isValid?: boolean) => void
};

const Label = styled.label`
  display: flex;
  flex-direction: column;
  margin-bottom: 0.75rem;

  > textarea {
    border: solid 1px #e4e4e4;
    border-radius: 5px;
    padding: 0.5em;
    margin: 0.5rem 0;
    font-size: 1rem;
    resize: vertical;
  }
`

class Textarea extends Component<TextareaProps, {}> {
  constructor(props: TextareaProps) {
    super(props);
  };

  componentDidMount = () => this.props.setValidationState(this.props.name);

  onChange = (event: ChangeEvent<HTMLTextAreaElement>) => {
    event.preventDefault();
    this.props.setNewFormState(this.props.name, event.currentTarget.value);
  };

  render = () => (
    <Label htmlFor={this.props.name}>
      {this.props.label}
      <textarea
        id={this.props.name}
        name={this.props.name}
        rows={5}
        className={this.props.validFields[this.props.name] === false ? 'red' : undefined}
        value={this.props.form[this.props.name] ? this.props.form[this.props.name] : ''}
        onChange={event => this.onChange(event)}
        onBlur={event => this.props.setValidationState(
          this.props.name,
          this.props.validator(event.currentTarget.value)
        )}
      />
    </Label>
  );
};

export default Textarea;