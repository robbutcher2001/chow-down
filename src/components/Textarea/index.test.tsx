import React from 'react';
import renderer from 'react-test-renderer';

import Textarea from '.';

const mockFormState = {
  field: 'value'
};

test('Textarea basic snapshot render', () => {
  const textarea = renderer.create(
    <Textarea
      name='title'
      label='The Textarea Title'
      form={mockFormState}
    />
  );

  expect(textarea.toJSON()).toMatchSnapshot();
});