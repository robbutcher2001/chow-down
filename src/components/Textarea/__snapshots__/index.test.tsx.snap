// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea basic snapshot render 1`] = `
<label
  className="sc-bdVaJa gtRYUw"
  htmlFor="title"
>
  The Textarea Title
  <textarea
    id="title"
    name="title"
    onBlur={[Function]}
    onChange={[Function]}
    rows={5}
    value=""
  />
</label>
`;
